app/app.go

```diff
@@ -929,6 +930,98 @@ func (app *App) FinalizeBlocker(ctx sdk.Context, req *abci.RequestFinalizeBlock)
 	return &resp, nil
 }
 
+func (app *App) DeliverTxWithResult(ctx sdk.Context, tx []byte) *abci.ExecTxResult {
+	deliverTxResp := app.DeliverTx(ctx, abci.RequestDeliverTx{
+		Tx: tx,
+	})
+	return &abci.ExecTxResult{
+		Code:      deliverTxResp.Code,
+		Data:      deliverTxResp.Data,
+		Log:       deliverTxResp.Log,
+		Info:      deliverTxResp.Info,
+		GasWanted: deliverTxResp.GasWanted,
+		GasUsed:   deliverTxResp.GasUsed,
+		Events:    deliverTxResp.Events,
+		Codespace: deliverTxResp.Codespace,
+	}
+}
+
+func (app *App) ProcessBlockSynchronous(ctx sdk.Context, txs [][]byte) []*abci.ExecTxResult {
+	txResults := []*abci.ExecTxResult{}
+	for _, tx := range txs {
+		txResults = append(txResults, app.DeliverTxWithResult(ctx, tx))
+	}
+	return txResults
+}
+
+// Returns a mapping of the access index to the channels
+func getChannelsFromSignalMapping(signalMapping map[acltypes.AccessOperation][]CompletionSignal) (channelsMapping [][](chan interface{})) {
+	for _, completionSignals := range signalMapping {
+		var channels []chan interface{}
+		for _, completionSignal := range completionSignals {
+			channels = append(channels, completionSignal.Channel)
+		}
+		channelsMapping = append(channelsMapping, channels)
+	}
+	return channelsMapping
+}
+
+func (app *App) ProcessTxConcurrent(
+	ctx sdk.Context,
+	txIndex int,
+	txBytes []byte,
+	wg *sync.WaitGroup,
+	resultChan chan<- *abci.ExecTxResult,
+	txCompletionSignalingMap map[acltypes.AccessOperation][]CompletionSignal,
+	txBlockingSignalsMap map[acltypes.AccessOperation][]CompletionSignal,
+) {
+	defer wg.Done()
+	// Store the Channels in the Context Object for each transaction
+	ctx.WithTxBlockingChannels(getChannelsFromSignalMapping(txBlockingSignalsMap))
+	ctx.WithTxCompletionChannels(getChannelsFromSignalMapping(txCompletionSignalingMap))
+
```

Comment:
can we add TODOs to block on blocking channels and writing to completion channels to properly block and release resources

Reply:
added here: https://github.com/sei-protocol/sei-chain/pull/261/files#diff-0f1d2976054440336a576d47a44a37b80cdf6701dd9113012bce0e3c425819b7R968